instrumentation/instagin/handler_test.go

```diff
@@ -0,0 +1,173 @@
+// (c) Copyright IBM Corp. 2021
+// (c) Copyright Instana Inc. 2016
+
+// +build go1.11
+
+package instagin_test
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"reflect"
+	"testing"
+
+	"github.com/instana/testify/require"
+
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/instana/testify/assert"
+
+	"github.com/gin-gonic/gin"
+	instana "github.com/instana/go-sensor"
+	"github.com/instana/go-sensor/instrumentation/instagin"
+)
+
+func TestMain(m *testing.M) {
+	gin.SetMode(gin.TestMode)
+	gin.DefaultWriter = ioutil.Discard
+
+	instana.InitSensor(&instana.Options{
+		Service: "gin-test",
+		Tracer: instana.TracerOptions{
+			CollectableHTTPHeaders: []string{"x-custom-header-1", "x-custom-header-2"},
+		},
+	})
+
+	os.Exit(m.Run())
+}
+
+func TestAddMiddleware(t *testing.T) {
+	const expectedHandlersAmount = 3
+	sensor := instana.NewSensor("gin-test")
+	engine := gin.Default()
+
+	handlerN1Pointer := reflect.ValueOf(engine.Handlers[0]).Pointer()
+	handlerN2Pointer := reflect.ValueOf(engine.Handlers[1]).Pointer()
+	assert.Len(t, engine.Handlers, expectedHandlersAmount-1)
+
+	instagin.AddMiddleware(sensor, engine)
+
+	assert.Len(t, engine.Handlers, expectedHandlersAmount)
+
+	// check that middleware was added as a first handler
+	assert.NotEqual(t, handlerN1Pointer, reflect.ValueOf(engine.Handlers[0]).Pointer())
+	assert.Equal(t, handlerN1Pointer, reflect.ValueOf(engine.Handlers[1]).Pointer())
+	assert.Equal(t, handlerN2Pointer, reflect.ValueOf(engine.Handlers[2]).Pointer())
+}
+
+func TestAddMiddlewareIdempotence(t *testing.T) {
+	const expectedHandlersAmount = 3
+	sensor := instana.NewSensor("gin-test")
+	engine := gin.New()
+
+	instagin.AddMiddleware(sensor, engine)
+	engine.Use(gin.Logger(), gin.Recovery())
+
+	// add middleware second time
+	instagin.AddMiddleware(sensor, engine)
+
+	assert.Len(t, engine.Handlers, expectedHandlersAmount)
+}
+
+func TestResponseHeadersCollecting(t *testing.T) {
+	sensor := instana.NewSensor("gin-test")
+	engine := gin.Default()
+	instagin.AddMiddleware(sensor, engine)
+	engine.GET("/foo", func(c *gin.Context) {
+		c.JSON(200, gin.H{})
+	})
+
+	req := httptest.NewRequest("GET", "/foo", nil)
+	w := httptest.NewRecorder()
+
+	engine.ServeHTTP(w, req)
+
+	assert.NotEmpty(t, w.Header().Get("X-Instana-T"))
+	assert.NotEmpty(t, w.Header().Get("X-Instana-S"))
+	assert.NotEmpty(t, w.Header().Get("X-Instana-L"))
+	assert.NotEmpty(t, w.Header().Get("Traceparent"))
+	assert.NotEmpty(t, w.Header().Get("Tracestate"))
+}
+
+func TestPropagation(t *testing.T) {
+	traceIDHeader := "0000000000001234"
+	spanIDHeader := "0000000000004567"
+
+	recorder := instana.NewTestRecorder()
+	tracer := instana.NewTracerWithEverything(
+		&instana.Options{Tracer: instana.TracerOptions{
+			CollectableHTTPHeaders: []string{"x-custom-header-1", "x-custom-header-2"},
+		},
+		}, recorder)
+
+	sensor := instana.NewSensorWithTracer(tracer)
+
+	engine := gin.Default()
```

Comment:
A minor note: you might consider extracting a helper function that initializes an instrumented Gin engine.

Reply:
I was thinking about it and found a few ways to create an engine object. Users can create it with default parameters (method **Default**), without parameters (method **New**) or just create an object with &gin.Engine. Therefore my idea was to use a unified way to cover all these cases with **AddMiddleware**. In my opinion, it will simplify usage.

Reply:
Right, my idea was to make it a private test helper and not generalize it too much. But it's up to you!

Reply:
I read the wrong filename (did not recognize that it is about _test) 😄  I will check it once. Good catch 👍

Reply:
done in child PR